container/queue.go

```diff
@@ -19,108 +19,373 @@ package container
 
 import (
 	"errors"
+	"runtime"
 	"sync"
-)
-
-// minQueueLen is smallest capacity that queue may have.
-// Must be power of 2 for bitwise modulus: x % n == x & (n - 1).
-const (
-	defaultQueueLen = 16
+	"sync/atomic"
+	"time"
 )
 
 var (
-	ErrEmpty = errors.New("quue is empty")
-	ErrFull  = errors.New("queue is full")
+	// ErrDisposed is returned when an operation is performed on a disposed
+	// queue.
+	ErrDisposed = errors.New(`queue: disposed`)
+
+	// ErrTimeout is returned when an applicable queue operation times out.
+	ErrTimeout = errors.New(`queue: poll timed out`)
+
+	// ErrEmptyQueue is returned when an non-applicable queue operation was called
+	// due to the queue's empty item state
+	ErrEmptyQueue = errors.New(`queue: empty queue`)
 )
 
-// Queue represents a single instance of the queue data structure.
-type Queue struct {
-	buf                    []interface{}
-	head, tail, count, cap int
-	lock                   sync.Mutex
+type waiters []*sema
+
+func (w *waiters) get() *sema {
+	if len(*w) == 0 {
+		return nil
+	}
+
+	sema := (*w)[0]
+	copy((*w)[0:], (*w)[1:])
+	(*w)[len(*w)-1] = nil // or the zero value of T
+	*w = (*w)[:len(*w)-1]
+	return sema
 }
 
-// New constructs and returns a new Queue.
-func NewQueue(cap int) *Queue {
-	return &Queue{
-		cap: cap,
-		buf: make([]interface{}, defaultQueueLen),
+func (w *waiters) put(sema *sema) {
+	*w = append(*w, sema)
+}
+
+func (w *waiters) remove(sema *sema) {
+	if len(*w) == 0 {
+		return
+	}
+	// build new slice, copy all except sema
+	ws := *w
+	newWs := make(waiters, 0, len(*w))
```

Comment:
wtf. ugly implementation. u should not use another memory arena.

Reply:
done